tools/archives/esap_integration/astronomical_archives.py

```diff
@@ -0,0 +1,476 @@
+import sys
+import json
+import re
+
+from urllib import request, parse
+
+
+class ArchiveRequest:
+
+    def __init__(self, base_url, request_parameters):
+        self._base_url = base_url
+        self._request_parameters = dict(request_parameters)
+        self._request = ''
+
+    def create_request(self) -> str:
+
+        for key in list(self._request_parameters.keys()):
+            if self._request_parameters[key] == '':
+                del self._request_parameters[key]
+
+        querystring = parse.urlencode(self._request_parameters)
+
+        request_url = self._base_url + '?' + querystring
+
+        self._request = request_url
+
+        return request_url
+
+
+class Archive:
+
+    def __init__(self):
+        self._name = "archive"
+        self.base_url = ""
+
+    def query_archive1(self, base_url, url_params) -> None:
+        archive_querystring = parse.urlencode(url_params)
+
+        with request.urlopen(base_url + '?' + archive_querystring) as response:
+            file = response.read()
+
+    def query_archive_request(self, request) -> None:
+
+        with request.urlopen(request) as response:
+            file = response.read()
+
+
+class AstronArchive(Archive):
+
+    name = "Astron Archive"
+    archive_type_name = "astron"
+
+    archive_query_parameters = {
+        'level': 'processed',
+        'target': '',
+        'ra': '',
+        'dec': '',
+        'fov': '',
+        'collection': '',
+        'page': '',
+        'archive_uri': 'astron_vo'
+    }
+
+    def __init__(self):
+        super().__init__()
+        self._base_url = "https://sdc-dev.astron.nl/esap-api/query/query/"
```

Comment:
we really do not want to query ESAP. I talking about ESAP as an example of how a platform like galaxy can query IVOA archives. You may note that eventual archivess tool in galaxy looks a bit like ESAP IVOA query in functinality.

Reply:
I thought the goal was to integrate esap since it's  also a wp5 project. Since anyone can put some esap gateway online and add some specific archive I thought it would be good since it would make it possible to integrate more esap gateways with different archives that might not be accessible through ivoa (like cta if I'm not mistaken) also like with the astron archive it enables the archive to put more specific parameters (like collections) that I wasn't able to find with pyvo.


If you don't want esap we can use only the pyvo integration but what I had in mind was that the two could be merged to either work as fallback if no esap gateway are available or to let the user choose if he wants to query an esap gateway or use pyvo directly.

Reply:
> I thought the goal was to integrate esap since it's also a wp5 project. Since anyone can put some esap gateway online and add some specific archive I thought it would be good since it would make it possible to integrate more esap gateways with different archives that might not be accessible through ivoa (like cta if I'm not mistaken) also like with the astron archive it enables the archive to put more specific parameters (like collections) that I wasn't able to find with pyvo.

There is not really a guarantee that there will be many ESAP interfaces. 
The baseline interface which is common, and will be adopted in some way across more current project of ours, is the IVOA interface (interfaces).

If pyvo does not support some necessary part of IVOA protocols, you can try to implement it yourself in python following specification. Or even propose a change to pyvo.
If esap supports something which is not supported by IVOA, then we can not afford to rely on such a feature since there is no guarantee that this feature will remain supported. So missing functionality should be added in the client tool.

> 
> If you don't want esap we can use only the pyvo integration but what I had in mind was that the two could be merged to either work as fallback if no esap gateway are available or to let the user choose if he wants to query an esap gateway or use pyvo directly.

Relying on ESAP will have peculiar and in most cases unfortunate implications of all kinds, most of all with sustainability. Since you already implemented something which relies on ESAP, I suggest you move it to a separate tool and not put much effort on it anymore. We probably could use this experience, but it's something very different.

Meanwhile, we really need a tool which works on IVOA protocols directly, like pyvo does. Note that there are also  IVOA services not in this registry you see. You can use pyvo when suitable, but most of all please rely on protocols.

Reply:
Ok I will switch to the pyvo version and try to depend as less as possible on pyvo and more on directly interacting with the protocol. I didn't thought you wanted some from scratch implementation since it represents a good part of the pyvo codebase but I think it could be done. I tried that way at first so I have some ideas on how it could work, will work on it this weekend so I can get back to you on that

Reply:
Well whatever pyvo provides is good to use. If it turns out that it misses some important features we can propose to add them actually, it will be much appreciated by the community.

Certainly there is no need to reimplement pyvo, just extend it.

Unless it turns out that pyvo is not suitable at all, but I doubt it.